web/src/contexts/BlockContext.tsx

```diff
@@ -0,0 +1,93 @@
+import { ReactNode, createContext, useContext, useState } from 'react'
+
+import { create } from '@bufbuild/protobuf'
+
+import {
+  Block,
+  BlockKind,
+  BlockRole,
+  BlockSchema,
+  GenerateRequest,
+  GenerateRequestSchema,
+} from '../gen/es/cassie/blocks_pb'
+import { useClient as useAgentClient } from './AgentContext'
+
+type BlockContextType = {
+  blocks: Block[]
+  sendUserBlock: (text: string) => Promise<void>
+  isInputDisabled: boolean
+}
+
+const BlockContext = createContext<BlockContextType | undefined>(undefined)
+
+// eslint-disable-next-line react-refresh/only-export-components
+export const useBlock = () => {
+  const context = useContext(BlockContext)
+  if (!context) {
+    throw new Error('useBlock must be used within a BlockProvider')
+  }
+  return context
+}
+
+export const BlockProvider = ({ children }: { children: ReactNode }) => {
+  const { client } = useAgentClient()
+  const [blocks, setBlocks] = useState<Block[]>([])
+  const [isInputDisabled, setIsInputDisabled] = useState(false)
+
+  const sendUserBlock = async (text: string) => {
+    if (!text.trim()) return
+
+    const userBlock = create(BlockSchema, {
+      role: BlockRole.USER,
+      kind: BlockKind.MARKUP,
+      contents: text,
+    })
+
+    const assistantBlock = create(BlockSchema, {
+      role: BlockRole.ASSISTANT,
+      kind: BlockKind.MARKUP,
+      contents: '...',
+    })
+
+    setBlocks((prevBlocks) => [...prevBlocks, userBlock, assistantBlock])
+    setIsInputDisabled(true)
+
+    const req: GenerateRequest = create(GenerateRequestSchema, {
+      blocks: [userBlock],
+    })
+
+    try {
+      const res = client!.generate(req)
+      for await (const r of res) {
+        const block = r.blocks[r.blocks.length - 1]
+        setBlocks((prevBlocks) => {
```

Comment:
I think this implementation might be making some assumptions about the protocol used by the backend that aren't accurate.

Here are the key ideas. 
1. a Generate request will return a stream of GenerateResponse
2. Each GenerateResponse could contain multiple blocks
3. Each block will be assigned a unique stable ID
4. Subsequent GenerateResponse could contain updates to a previously returned block.
    * In this case the block id will match the id of a previously returned block
    * The block returned in the GenerateResponse will contain a complete copy of the block so the client can just override the block with the newly returned block

Reply:
Here's my previous attempt
https://github.com/jlewi/cloud-assistant/blob/bf86a917dd792f4b1b6154c95388d79828304346/ui/app/components/blocks-context.tsx#L37